File: midterm2/q09/fibo.scala
import scala.annotation.tailrec

object Fibonacci {
    def main(args: Array[String]) = {
        println(fibonacci(400))
    }

    def fibonacci(position : Int) : BigInt = position match{
        case 0 => BigInt(1)
        case 1 => BigInt(1)
        case _ => {
            fibonacciHelper(2, position, List(1, 1))
        }
    }

    @tailrec final def fibonacciHelper(position : Int, upperLimit : Int, lookUpTable : List[BigInt]) : BigInt = {
        if(position < upperLimit) {
            val currentValue = lookUpTable.apply(position - 2) + lookUpTable.apply(position - 1)
            fibonacciHelper(position + 1, upperLimit, lookUpTable :+ currentValue)
        }else {
            lookUpTable.apply(position - 2) + lookUpTable.apply(position - 1)
        }
    }
}

File: midterm2/q06/fibo.rb
RubyVM::InstructionSequence.compile_option = {
    tailcall_optimization: true,
    trace_instruction: false
}

def fibonacci(position)
    case position    
    when 0
        0
    when 1
        0
    else
        fibonacciHelper(2, position, [1, 1])
    end
end

def fibonacciHelper(position, upperLimit, lookUpTable)
    if(position < upperLimit)
        currentValue = lookUpTable[position - 2] + lookUpTable[position - 1]
        fibonacciHelper(position + 1, upperLimit, lookUpTable + [currentValue])
    else
        lookUpTable[position - 2] + lookUpTable[position - 1]
    end
end

puts fibonacci(400)
